pkg/smartcontract/entry.go

```diff
@@ -20,6 +20,60 @@ import (
 // for interactions with RPC server that have iterator sessions disabled.
 func CreateCallAndUnwrapIteratorScript(contract util.Uint160, operation string, maxIteratorResultItems int, params ...any) ([]byte, error) {
 	script := io.NewBufBinWriter()
+	jmpIfNotOffset, jmpIfMaxReachedOffset := emitCallAndUnwrapIteratorScript(script, contract, operation, maxIteratorResultItems, params...)
+
+	// End of the program: push the result on stack and return.
+	loadResultOffset := script.Len()
+	emit.Opcodes(script.BinWriter, opcode.NIP, // Remove iterator from the 1-st cell of estack
+		opcode.NIP) // Remove maxIteratorResultItems from the 1-st cell of estack, so that only resulting array is left on estack.
+	if err := script.Err; err != nil {
+		return nil, fmt.Errorf("emitting iterator unwrapper script: %w", err)
+	}
+
+	// Fill in JMPIFNOT instruction parameter.
+	bytes := script.Bytes()
+	bytes[jmpIfNotOffset+1] = uint8(loadResultOffset - jmpIfNotOffset) // +1 is for JMPIFNOT itself; offset is relative to JMPIFNOT position.
+	// Fill in jmpIfMaxReachedOffset instruction parameter.
+	bytes[jmpIfMaxReachedOffset+1] = uint8(loadResultOffset - jmpIfMaxReachedOffset) // +1 is for JMPIF itself; offset is relative to JMPIF position.
+	return bytes, nil
+}
+
+// CreateCallAndPrefetchIteratorScript creates a script that calls 'operation' method
+// of the 'contract' with the specified arguments. This method is expected to return
+// an array of the first iterator items (up to maxIteratorResultItems, which cannot exceed VM limits)
+// and, optionally, an iterator that then is traversed (using iterator.Next).
+// The result of the script is an array containing extracted value elements and an iterator, if it can contain more items.
+// Note, however, that if an iterator is returned, the number of remaining items can still be 0.
+// This script should only be used for interactions with RPC server that have iterator sessions enabled.
+func CreateCallAndPrefetchIteratorScript(contract util.Uint160, operation string, maxIteratorResultItems int, params ...any) ([]byte, error) {
+	script := io.NewBufBinWriter()
+	jmpIfNotOffset, jmpIfMaxReachedOffset := emitCallAndUnwrapIteratorScript(script, contract, operation, maxIteratorResultItems, params...)
+
+	// 1st possibility: jump here when the maximum number of items was reached.
+	retainIteratorOffset := script.Len()
+	emit.Opcodes(script.BinWriter, opcode.ROT, // Put maxIteratorResultItems from the 2-nd cell of estack, to the top
+		opcode.DROP, // ... and then drop it.
+		opcode.SWAP, // Put the iterator on top of the stack.
+		opcode.RET)
+
+	// 2nd possibility: jump here when the iterator has no more items.
+	loadResultOffset := script.Len()
+	emit.Opcodes(script.BinWriter, opcode.ROT, // Put maxIteratorResultItems from the 2-nd cell of estack, to the top
+		opcode.DROP, // ... and then drop it.
+		opcode.NIP)  // Drop iterator as the 1-st cell on the stack.
```

Comment:
This is an end portion of the script, which is a target for an earlier jump. How do you suggest uniting them?

Reply:
Actually, we can avoid jumps now, because 2 exit conditions are different! This will save us a couple instructions at the cost of inability to reuse the actual traversing part of a script. I find code reuse here much more useful, bytecode is hard to review.